### PR TITLE
use goroutines

### PR DESCRIPTION
**Pull request by Kishan-Dhakan**
### Changes
-

### Fixes
-

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/gosdk/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- blobber:
- 0chain:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: zboxcore/zboxutil/http.go [hunk 1]
```diff
@@ -534,38 +535,35 @@ func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]
...
-				retObj = entityBytes
+			urlObj.RawQuery = q.Encode()
+			client := &http.Client{Transport: DefaultTransport}
+
+			response, err := client.Get(urlObj.String())
+			if err == nil {
+				defer response.Body.Close()
+				entityBytes, _ := ioutil.ReadAll(response.Body)
+
+				responses[response.StatusCode]++
```
Comment (peterlimg) [inline zboxcore/zboxutil/http.go:542]: The `responses` slice is not thread safe

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: zboxcore/zboxutil/http.go (hunk 1 above)
Comment (Kishan-Dhakan) [reply zboxcore/zboxutil/http.go:542]: Added mutex locks to prevent race conditions.

File: zboxcore/zboxutil/http.go [hunk 2]
```diff
@@ -531,41 +532,41 @@ func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]
...
+				entityBytes, _ := ioutil.ReadAll(response.Body)
+
+				mu.Lock()
+				responses[response.StatusCode]++
+				mu.Unlock()
+				if responses[response.StatusCode] > maxCount {
+					maxCount = responses[response.StatusCode]
+					retObj = entityBytes
+				}
+				entityResult[sharder] = retObj
```
Comment (cnlangzi) [inline zboxcore/zboxutil/http.go:547]: reading responses and entityResult are also non-thread-safe here. mu.Unlock should be moved to line 562

**Review by cnlangzi — CHANGES_REQUESTED** [review]

File: zboxcore/zboxutil/http.go (hunk 2 above)
Comment (Kishan-Dhakan) [reply zboxcore/zboxutil/http.go:547]: Nice catch Dayi!

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | master  |
| gosdk  | enhance/zboxcoreRESTcall  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: zboxcore/zboxutil/http.go (hunk 2 above)
Comment (cnlangzi) [reply zboxcore/zboxutil/http.go:547]:
```suggestion
entityResult[sharder] = retObj
mu.Unlock()
```

**Review by cnlangzi — CHANGES_REQUESTED** [review]

**Review by cnlangzi — APPROVED** [review]

**Review by peterlimg — APPROVED** [review]
lgtm

**Pull request merged**